advanced_server.rb

```diff
@@ -0,0 +1,314 @@
+#
+# On its own, this app does absolutely nothing, except that it can be installed.
+# It's up to you to add fun functionality!
+# You can check out one example in advanced_server.rb.
+#
+# This code is a Sinatra app, for two reasons:
+#   1. Because the app will require a landing page for installation.
+#   2. To easily handle webhook events.
+#
+#
+# Of course, not all apps need to receive and process events!
+# Feel free to rip out the event handling code if you don't need it.
+#
+# Have fun!
+#
+
+class GHAapp < Sinatra::Application
+
+  # !!! DO NOT EVER USE HARD-CODED VALUES IN A REAL APP !!!
+  # Instead, set and read app tokens or other secrets in your code
+  # in a runtime source, like an environment variable like below
+
+  # Expects that the private key has been set as an environment variable in
+  # PEM format using the following command to replace newlines with the
+  # literal `\n`:
+  #   export GITHUB_PRIVATE_KEY=`awk '{printf "%s\\n", $0}' private-key.pem`
+  #
+  # Converts the newlines
+  PRIVATE_KEY = OpenSSL::PKey::RSA.new(ENV['GITHUB_PRIVATE_KEY'].gsub('\n', "\n"))
+
+  # Your registered app must have a secret set. The secret is used to verify
+  # that webhooks are sent by GitHub.
+  WEBHOOK_SECRET = ENV['GITHUB_WEBHOOK_SECRET']
+
+  # The GitHub App's identifier (type integer) set when registering an app.
+  APP_IDENTIFIER = ENV['GITHUB_APP_IDENTIFIER']
+
+  # Turn on Sinatra's verbose logging during development
+  configure :development do
+    set :logging, Logger::DEBUG
+  end
+
+
+  # Before each request to the `/event_handler` route
+  before '/event_handler' do
+    get_payload_request(request)
+    verify_webhook_signature
+    authenticate_app
+    # Authenticate each installation of the app in order to run API operations
+    authenticate_installation(@payload)
+  end
+
+
+  post '/event_handler' do
+    # Get the event type from the HTTP_X_GITHUB_EVENT header
+    case request.env['HTTP_X_GITHUB_EVENT']
+
+    when 'check_suite'
+      # A new check_suite has been created. Create a new check run with status queued
+      if @payload['action'] === 'requested' || @payload['action'] === 'rerequested'
+        create_check_run
+      end
+
+    when 'check_run'
+      # GH confirms our new check_run has been created, or rerequested. Update it to "completed"
+      # Notice that we get notifications of the check runs created by _other_ systems than ours!
+      # We need to be selective, hence the conditional on the app id. We only want to process our _own_
+      # check runs. That's why we check if the app id is == APP_IDENTIFIER
+      if @payload['check_run']['app']['id'].to_s === APP_IDENTIFIER
+        case @payload['action']
+        when 'created'
+          initiate_check_run
+        when 'rerequested'
+          create_check_run
+        when 'requested_action'
+          take_requested_action
+        end
+      end
+    end
+  end
```

Comment:
I believe I needed to return some value to the POST request, `200` or `201`, to resolve an issue trying to return the check_run response to the a client.

Reply:
I added a `status 200` to be returned from the POST request. 👍